### DISCUSSION TITLE
autocommit=True makes no effect when object changed

### DISCUSSION THREAD

**Discussion by scruel** [sqlalchemy/sqlalchemy#6927] [Usage Questions]
### Describe the bug

When set autocommit=True, the change of an object(instance) which got by query function will not commit to the database.
But when set autocommit=False, change object then execute commit works as excepted.

### To Reproduce

```python
engine = create_engine('mysql+pymysql://xx')

# set autocommit=True
session_factory = sessionmaker(class_=Session, bind=engine, autocommit=True)
session = session_factory()

t = session.query(Test).get(1)
# change the object
t.test = "111"
# no effect for now.
# execute commit function will cause InvalidRequestError
# session.commit()

session.merge(t)
# only merge function can take effect.
# if autocommit=False, commit will take effect.
```

### Error

sqlalchemy.exc.InvalidRequestError: No transaction is begun.

### Versions

- OS: Win10
- Python: 3.6
- SQLAlchemy: 1.1.13
- Database: mysql

**Comment by scruel** [comment]
With version 1.4.23, the problem still exists.

When set autoflush=True, it still needs to execute flush function manually even the object already changed.
```
t = session.query(Test).get(1)
t.test = "111"
# no effect for now.

session.flush()
# taken effect.

# session.commit()
```

**Comment by scruel** [comment]
By the way, is there any argument I can use to stop bind the model object with session actions? I just want the model object as a model object, totally confused about the design of update action in sqlalchemy.

**Comment by CaselIT** [comment]
Hi,

I think this is has always beed the expected behaviour. Autocommit is a deprecated feature, so I don't think anything will change in this regard.

> By the way, is there any argument I can use to stop bind the model object with session actions? I just want the model object as a model object, totally confused about the design of update action in sqlalchemy.

not sure I understand what you mean. Maybe `session.expunge` is what you seek?

**Comment by scruel** [comment]
> 
> I think this is has always beed the expected behaviour. Autocommit is a deprecated feature, so I don't think anything will change in this regard.

So I have to remember to commit after invoke any sqlalchemy function?

> 
> not sure I understand what you mean. Maybe `session.expunge` is what you seek?
I was a Java developer and familiar with Mybatis, It could just execute like this:
```
object = Test();
object.setId(1);
object.setXXX(xxx);
xxxDao.updateByKey(object);
```

And any change of object will not cause the change of database, the model object is just a POJO, not like sqlalchemy will bind session instance to it.

`session.expunge` still not work as expected:

```
t = session.query(Test).get(1)
session.expunge(t)
t.test = "111"
session.flush()
session.merge(t)
session.commit()
```
No effect will cause by the above code(with autocommit=True method).

However, as you mentioned that "Autocommit is a deprecated feature", so when set autocommit=False, it seems to work fine, but I want to unbind all objects with sqlalchemy, not use `session.expunge(t)` to unbind just a single one.

**Comment by CaselIT** [comment]
> So I have to remember to commit after invoke any sqlalchemy function?

As documented here https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.params.autocommit you need to flush or commit for the change to be saved in the db

> `session.expunge` still not work as expected:
> 
> ```
> t = session.query(Test).get(1)
> session.expunge(t)
> t.test = "111"
> session.flush()
> session.merge(t)
> session.commit()
> ```

Expunge there doesn't make sense, since you want to update the object. Expunge is used when you want to deassociste an object from the session, like when persisting it in a cache of some kind.

> However, as you mentioned that "Autocommit is a deprecated feature", so when set autocommit=False, it seems to work fine, but I want to unbind all objects with sqlalchemy, not use `session.expunge(t)` to unbind just a single one.

There is `expunge_all`. I still don't understand what you are trying to do. Can you provide an example?

**Comment by scruel** [comment]
> There is `expunge_all`. I still don't understand what you are trying to do. Can you provide an example?

What I want is a pure POJO when get from the session, rather than an object with session instance which would be so coupled.
When modify the object, it's just modified the object, rather than flush the change to DB, and the autoFlush/autoCommit behavior seems not predictable, so I may can't know what will happen when I change the object after update or something, it's not a good design...
`expunge_all` is but a function but not an engine parameter, so I need to invoke it after each get from the session.

**Comment by gordthompson** [comment]
> So I have to remember to commit after invoke any sqlalchemy function?

No. As explained [here](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block), you can use a context manager to automatically  commit unless an error occurs, in which case it will automatically roll back.

```python
with Session(engine) as session, session.begin():
    thing1 = session.get(Thing, 1)
    thing1.txt = "new text"
```

The above change will automatically be committed provided that no error was raised.

If you want to use a Session to retrieve some ORM objects and then work with them as "pure POJO [sic]" objects then AFAIK you'll need to decouple them (i.e., expunge them from the session) yourself. That just how ORM works.
